File: ContestOfSwift/TestsOfSwift/Challenges/Challenge8Tests.swift
//
//  Challenge8Tests.swift
//  TestsOfSwift
//
//  Created by Sergio Daniel on 4/29/24.
//

import XCTest

final class Challenge8Tests: XCTestCase {

    func testCountInstances_withMultipleOccurrences() {
        let words = ["apple", "banana", "Apple", "orange", "aPPle"]
        let count = Challenge8.countInstances(of: "apple", in: words)
        XCTAssertEqual(count, 3, "The count of 'apple' should be 3, ignoring case.")
    }

    func testCountInstances_withNoOccurrences() {
        let words = ["apple", "banana", "grape"]
        let count = Challenge8.countInstances(of: "orange", in: words)
        XCTAssertEqual(count, 0, "The count of 'orange' should be 0.")
    }

    func testCountInstances_caseInsensitivity() {
        let words = ["apple", "Apple", "APPLE"]
        let count = Challenge8.countInstances(of: "apple", in: words)
        XCTAssertEqual(count, 3, "The count of 'apple' should be 3, ignoring case differences.")
    }

    func testCountInstances_emptyArray() {
        let words: [String] = []
        let count = Challenge8.countInstances(of: "apple", in: words)
        XCTAssertEqual(count, 0, "The count should be 0 for an empty array.")
    }

    func testCountInstances_emptyString() {
        let words = ["apple", "banana", "", "", "grape"]
        let count = Challenge8.countInstances(of: "", in: words)
        XCTAssertEqual(count, 0, "The count of an empty string should be 0.")
    }

    func testCountInstances_specialCharacters() {
        let words = ["apple!", "banana?", "Apple!", "orange."]
        let count = Challenge8.countInstances(of: "apple", in: words)
        XCTAssertEqual(count, 0, "The count of 'apple' should be 0, ignoring case.")
    }

    func testCountWords_basicFunctionality() {
        let text = "Hello, hello? HELLO! Are there, are there any echoes?"
        let counts = Challenge8.countWords(in: text)
        XCTAssertEqual(counts, ["hello": 3, "are": 2, "there": 2, "any": 1, "echoes": 1], "Word counts should match expected values, ignoring case and punctuation.")
    }

    func testCountWords_emptyString() {
        let text = ""
        let counts = Challenge8.countWords(in: text)
        XCTAssertTrue(counts.isEmpty, "The result should be an empty dictionary for an empty string.")
    }

    func testCountWords_caseInsensitivity() {
        let text = "Swift, swift, SWIFT!"
        let counts = Challenge8.countWords(in: text)
        XCTAssertEqual(counts, ["swift": 3], "The count of 'swift' should be 3, ignoring case.")
    }

    func testCountWords_ignorePunctuation() {
        let text = "Well... well, well; well!"
        let counts = Challenge8.countWords(in: text)
        XCTAssertEqual(counts, ["well": 4], "The count of 'well' should be 4, ignoring all punctuation.")
    }

    func testCountWords_withNumbersAndSpecialCharacters() {
        let text = "Testing 123, testing 123! Does it count 123? Yes, it counts!"
        let counts = Challenge8.countWords(in: text)
        XCTAssertEqual(counts, ["testing": 2, "does": 1, "it": 2, "count": 1, "counts": 1, "yes": 1], "The count should ignore numbers and special characters, except for apostrophes in words.")
    }

    func testCountWords_singleWordRepetition() {
        let text = "Test test test test"
        let counts = Challenge8.countWords(in: text)
        XCTAssertEqual(counts, ["test": 4], "The count of 'test' should be 4.")
    }

}

File: ContestOfSharp/TestOfSharp/Challenges/Challenge8Tests.cs
using BankaiSouth;

namespace TestOfSharp.Challenges;

using static ContestOfSharp.Challenges.Challenge8;

[TestClass]
public class Challenge8Tests
{
    [TestMethod]
    public void TestCountInstances_WithMultipleOccurrences()
    {
        const string text = "apple banana Apple orange aPPle grape";
        var count = CountInstancesOf(word: "apple", inText: text);
        Assert.AreEqual(expected: 3, actual: count);
    }
    
    [TestMethod]
    public void TestCountInstances_WithNoOccurrences()
    {
        const string text = "apple banana Apple orange aPPle grape";
        var count = CountInstancesOf(word: "pear", inText: text);
        Assert.AreEqual(expected: 0, actual: count);
    }
    
    [TestMethod]
    public void TestCountInstances_WithEmptyText()
    {
        const string text = "";
        var count = CountInstancesOf(word: "apple", inText: text);
        Assert.AreEqual(expected: 0, actual: count);
    }
    
    [TestMethod]
    public void TestCountInstances_WithEmptyWord()
    {
        const string text = "apple banana Apple orange aPPle grape";
        var count = CountInstancesOf(word: "", inText: text);
        Assert.AreEqual(expected: 0, actual: count);
    }
    
    [TestMethod]
    public void TestCountInstances_WithEmptyTextAndWord()
    {
        const string text = "";
        var count = CountInstancesOf(word: "", inText: text);
        Assert.AreEqual(expected: 0, actual: count);
    }
    
    [TestMethod]
    public void TestCountInstances_WithSpecialCharacters()
    {
        const string text = "apple! banana? Apple! orange. grape";
        var count = CountInstancesOf(word: "apple", inText: text);
        Assert.AreEqual(expected: 2, actual: count);
    }

    [TestMethod]
    public void TestCountWordsIn_BasicFunctionality()
    {
        const string text = "Hello, hello? HELLO! Are there, are there any echoes?";
        CollectionAssert.AreEqual(
            expected: new Dictionary<string, int>
            {
                ["hello"] = 3, ["are"] = 2, ["there"] = 2, ["any"] = 1, ["echoes"] = 1
            },
            actual: CountWordsIn(text)
        );
    }
    
    [TestMethod]
    public void TestCountWordsIn_EmptyText()
    {
        CollectionAssert.AreEqual(
            expected: new Dictionary<string, int>(),
            actual: CountWordsIn("")
        );
    }
    
    [TestMethod]
    public void TestCountWordsIn_CaseInsensitivity()
    {
        CollectionAssert.AreEqual(
            expected: new Dictionary<string, int> { ["swift"] = 3 },
            actual: CountWordsIn("Swift, swift, SWIFT!")
        );
    }

    [TestMethod]
    public void TestCountWordsIn_IgnorePunctuation()
    {
        CollectionAssert.AreEqual(
            expected: new Dictionary<string, int> { ["well"] = 4 },
            actual: CountWordsIn("Well... well, well; well!")
        );
    }
    
    // create case for "Testing 123, testing 123! Does it count 123? Yes, it counts!"
    [TestMethod]
    public void TestCountWordsIn_WithNumbers()
    {
        var result = CountWordsIn("Testing 123, testing 123! Does it count 123? Yes, it counts!");
        CollectionAssert.AreEqual(
            expected: new Dictionary<string, int> { ["testing"] = 2, ["does"] = 1, ["it"] = 2, ["count"] = 1, ["yes"] = 1, ["counts"] = 1 },
            actual: result
        );
    }
}